File: app/src/main/java/xyz/volgoak/wordlearning/entities/Word.java
package xyz.volgoak.wordlearning.entities;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.PrimaryKey;

import com.google.gson.annotations.Expose;

import xyz.volgoak.wordlearning.data.DatabaseContract.Words;

/**
 * Created by alex on 1/3/18.
 */

@Entity(tableName = Words.TABLE_NAME)
public class Word implements DataEntity {

    @ColumnInfo(name = Words._ID)
    @PrimaryKey(autoGenerate = true)
    private long id;
    @Expose
    @ColumnInfo(name = Words.COLUMN_WORD)
    private String word;
    @Expose
    @ColumnInfo(name = Words.COLUMN_TRANSLATION)
    private String translation;
    @Expose
    @ColumnInfo(name = Words.COLUMN_TRANSCRIPTION)
    private String transcription;
    @ColumnInfo(name = Words.COLUMN_TRAINED_WT)
    private int trainedWt;
    @ColumnInfo(name = Words.COLUMN_TRAINED_TW)
    private int trainedTw;
    @ColumnInfo(name = Words.COLUMN_STUDIED)
    private int studied;
    @ColumnInfo(name = Words.COLUMN_STATUS)
    private int status;

    @Ignore
    public Word(String word) {
        this.word = word;
    }

    @Ignore
    public Word(String word, String translation) {
        this.word = word;
        this.translation = translation;
        this.status = Words.IN_DICTIONARY;
    }

    public Word(long id, String word, String translation, String transcription, int trainedWt,
                int trainedTw, int studied, int status) {
        this.id = id;
        this.word = word;
        this.translation = translation;
        this.transcription = transcription;
        this.trainedWt = trainedWt;
        this.trainedTw = trainedTw;
        this.studied = studied;
        this.status = status;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getWord() {
        return word;
    }

    public void setWord(String word) {
        this.word = word;
    }

    public String getTranslation() {
        return translation;
    }

    public void setTranslation(String translation) {
        this.translation = translation;
    }

    public String getTranscription() {
        return transcription;
    }

    public void setTranscription(String transcription) {
        this.transcription = transcription;
    }

    public int getTrainedWt() {
        return trainedWt;
    }

    public void setTrainedWt(int trainedWt) {
        this.trainedWt = trainedWt;
    }

    public int getTrainedTw() {
        return trainedTw;
    }

    public void setTrainedTw(int trainedTw) {
        this.trainedTw = trainedTw;
    }

    public int getStudied() {
        return studied;
    }

    public void setStudied(int studied) {
        this.studied = studied;
    }

    public int getStatus() {
        return status;
    }

    public void setStatus(int status) {
        this.status = status;
    }

    public void resetProgress() {
        trainedTw = 0;
        trainedWt = 0;
        studied = 0;
    }
}

File: data/src/main/java/com/attiladroid/data/entities/Word.kt
package com.attiladroid.data.entities

import android.arch.persistence.room.ColumnInfo
import android.arch.persistence.room.Entity
import android.arch.persistence.room.PrimaryKey
import com.attiladroid.data.DataContract.Words
import com.google.gson.annotations.Expose

/**
 * Created by alex on 1/3/18.
 */

@Entity(tableName = Words.TABLE_NAME)
data class Word(
        @ColumnInfo(name = Words._ID, index = true)
        @PrimaryKey(autoGenerate = true)
        override var id: Long = 0,
        @Expose
        @ColumnInfo(name = Words.COLUMN_WORD)
        var word: String = "",
        @Expose
        @ColumnInfo(name = Words.COLUMN_TRANSLATION)
        var translation: String = "",
        @Expose
        @ColumnInfo(name = Words.COLUMN_TRANSCRIPTION)
        var transcription: String? = null,
        @ColumnInfo(name = Words.COLUMN_TRAINED_WT)
        var trainedWt: Int = 0,
        @ColumnInfo(name = Words.COLUMN_TRAINED_TW)
        var trainedTw: Int = 0,
        @ColumnInfo(name = Words.COLUMN_STUDIED)
        var studied: Int = 0,
        @ColumnInfo(name = Words.COLUMN_STATUS)
        var status: Int = 0,
        var addedTime: Long = 0,
        var lastTrained: Long = 0)

    : DataEntity {

    val isInDictionary: Boolean
        get() = status == Words.IN_DICTIONARY

    fun resetProgress() {
        trainedTw = 0
        trainedWt = 0
        studied = 0
    }

    override fun toString(): String {
        return "$word, $translation, tw $trainedTw, wt $trainedWt"
    }
}
